### PR TITLE
executor: DeleteExec suport NextChunk

### PR DESCRIPTION
**Pull request by winkyao**
to #5261

### CODE REVIEW

File: executor/write.go [hunk 1]
```diff
@@ -197,6 +196,18 @@ func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
...
 
+// Next implements the Executor Next interface.
+func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
+	return e.exec(goCtx)
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *DeleteExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	_, err := e.exec(goCtx)
```
Comment (coocood) [inline executor/write.go]: Should avoid calling `SelectExec.Next` when `NextChunk` is called.

File: executor/builder.go [hunk 2]
```diff
@@ -863,13 +863,15 @@ func (b *executorBuilder) buildDelete(v *plan.Delete) Executor {
 		b.err = errors.Trace(b.err)
 		return nil
 	}
-	return &DeleteExec{
+	deleteExec := &DeleteExec{
 		baseExecutor: newBaseExecutor(nil, b.ctx),
```
Comment (coocood) [inline executor/builder.go]: make `selExec` a child.

**Comment by winkyao** [comment]
@coocood @zz-jason @XuHuaiyu PTAL

File: executor/write.go [hunk 3]
```diff
@@ -314,6 +298,149 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
...
+		err := e.children[0].NextChunk(goCtx, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		for rowIdx := 0; rowIdx < chk.NumRows(); rowIdx++ {
```
Comment (zz-jason) [inline executor/write.go]:
```go
for chunkRow := chk.Begin(); chunkRow != chk.End(); chunkRow = chunkRow.Next() {
}
```

File: executor/write.go [hunk 4]
```diff
@@ -314,6 +298,149 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
...
+		id, tbl = i, t
+		handleCol = e.children[0].Schema().TblID2Handle[id][0]
+		break
+	}
+
+	// If tidb_batch_delete is ON and not in a transaction, we could use BatchDelete mode.
+	batchDelete := e.ctx.GetSessionVars().BatchDelete && !e.ctx.GetSessionVars().InTxn()
+	fields := e.children[0].Schema().GetTypes()
+	for {
+		chk := chunk.NewChunk(fields)
```
Comment (zz-jason) [inline executor/write.go]:
```go
chk := e.children[0].newChunk()
```

File: executor/write.go [hunk 5]
```diff
@@ -314,6 +298,149 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
...
+	if len(e.Tables) == 0 {
+		return nil
+	}
+
+	e.initialMultiTableTblMap()
+	colPosInfos := e.getColPosInfos(e.children[0].Schema())
+	tblRowMap := make(tableRowMapType)
+	fields := e.children[0].Schema().GetTypes()
+	for {
+		chk := chunk.NewChunk(fields)
```
Comment (zz-jason) [inline executor/write.go]: ditto

File: executor/write.go [hunk 6]
```diff
@@ -314,6 +298,149 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
...
+		err := e.children[0].NextChunk(goCtx, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		for rowIdx := 0; rowIdx < chk.NumRows(); rowIdx++ {
```
Comment (zz-jason) [inline executor/write.go]: ditto

File: executor/builder.go [hunk 7]
```diff
@@ -901,13 +901,15 @@ func (b *executorBuilder) buildDelete(v *plan.Delete) Executor {
 		b.err = errors.Trace(b.err)
 		return nil
 	}
-	return &DeleteExec{
-		baseExecutor: newBaseExecutor(nil, b.ctx),
+	deleteExec := &DeleteExec{
```
Comment (zz-jason) [inline executor/builder.go]: can we un-export `SelectExec`, `Tables` and `IsMultiTable` ?

Comment (winkyao) [reply executor/builder.go]: we should modify it in another pr.

**Comment by winkyao** [comment]
@coocood @zz-jason PTAL

File: executor/write.go [hunk 8]
```diff
@@ -197,38 +197,40 @@ func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
 	return nil, e.deleteSingleTable(goCtx)
 }
 
+// NextChunk implements the Executor NextChunk interface.
+func (e *DeleteExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.finished {
+		return nil
+	}
+	defer func() {
```
[the remark this answers is not in the document]
Comment (winkyao) [reply executor/write.go]: e.finished must set after the exec finished, so use defer is more reasonable.

File: executor/write.go [hunk 9]
```diff
@@ -315,6 +299,148 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
...
+		err := e.children[0].NextChunk(goCtx, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		for joinedChunkRow := chk.Begin(); joinedChunkRow != chk.End(); joinedChunkRow = joinedChunkRow.Next() {
```
Comment (zz-jason) [inline executor/write.go]: what does `joinedChunkRow` mean ?

File: executor/write.go [hunk 10]
```diff
@@ -197,38 +197,40 @@ func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
...
+
 type tblColPosInfo struct {
 	tblID         int64
 	colBeginIndex int
 	colEndIndex   int
 	handleIndex   int
 }
 
+// tableRowMapType is a map for unique (Table, Row) pair.
+type tableRowMapType map[int64]map[int64][]types.Datum
```
Comment (zz-jason) [inline executor/write.go]: add a comment to explain what the key(`int64`) and value(`map[int64][]types.Datum`) represent for.

File: executor/write.go [hunk 11]
```diff
@@ -197,38 +197,40 @@ func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
...
+	chk.Reset()
+	if e.finished {
+		return nil
+	}
+	defer func() {
+		e.finished = true
+	}()
+
+	if e.IsMultiTable {
+		return e.deleteMultiTablesByChunk(goCtx)
```
Comment (zz-jason) [inline executor/write.go]: errors.Trace(e.deleteMultiTablesByChunk(goCtx))

File: executor/write.go [hunk 12]
```diff
@@ -197,38 +197,40 @@ func (e *DeleteExec) Next(goCtx goctx.Context) (Row, error) {
...
+	}
+	return e.deleteSingleTableByChunk(goCtx)
```
Comment (zz-jason) [inline executor/write.go]: ditto

File: executor/write.go (hunk 9 above)
Comment (winkyao) [reply executor/write.go]: delete multi tables actually delete the data from joined tables.

**Comment by zz-jason** [comment]
@coocood PTAL

**Pull request merged**